IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/ApiProtocol.swift

```diff
@@ -0,0 +1,89 @@
+//
+//  ApiProtocol.swift
+//  SwiftSeedProject
+//
+//  Created by Brian Sztamfater on 3/4/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import Foundation
+import Alamofire
+import SwiftyJSON
+
+protocol ApiManagerProtocol {
+    func apiRequest(endpoint: Endpoint, parameters: [String : AnyObject]?, headers: [String : String]?) -> ApiRequestProtocol
+}
+
+extension ApiManagerProtocol {
+    func apiRequest(endpoint: Endpoint) -> ApiRequestProtocol {
+        return apiRequest(endpoint: endpoint, parameters: nil, headers: nil)
+    }
+    
+    func apiRequest(endpoint: Endpoint, parameters: [String : AnyObject]?) -> ApiRequestProtocol {
+        return apiRequest(endpoint: endpoint, parameters: parameters, headers: nil)
+    }
+}
+
+protocol ApiRequestProtocol {
+    func apiResponse(completionHandler: @escaping (DataResponse<JSON>) -> Void) -> Self
+}
+
+func += <K, V> ( left: inout [K : V], right: [K : V]) {
+    for (k, v) in right {
+        left.updateValue(v, forKey: k)
+    }
+}
+
+extension SessionManager: ApiManagerProtocol {
+    func apiRequest(endpoint: Endpoint, parameters: [String : AnyObject]? = nil, headers: [String : String]? = nil) -> ApiRequestProtocol {
+        // Insert your common headers here, for example, authorization token or accept.
+        var commonHeaders = ["Accept" : "application/json"]
+        if let headers = headers {
+            commonHeaders += headers
+        }
+        
+        return request(endpoint.url, method: endpoint.method, parameters: parameters, encoding: JSONEncoding.default, headers: commonHeaders)
+    }
+}
+
+extension DataRequest: ApiRequestProtocol {
+    static func apiResponseSerializer() -> DataResponseSerializer<JSON> {
+        return DataResponseSerializer<JSON> { request, response, data, error in
+            if let error = error {
+                return .failure(error)
+            }
+            
+            guard let validData = data else {
+                let domain = Bundle.main.infoDictionary?[kCFBundleNameKey as String] as? String
+                let reason = "Data could not be serialized. Input data was nil."
+                return .failure(NSError(domain: domain!, code: 1001, userInfo: [NSLocalizedDescriptionKey : reason]))
```

Comment:
![asdasdasdsadsa](https://cloud.githubusercontent.com/assets/6931639/24733012/36a09ff4-1a4b-11e7-88c3-d500f46451e0.png)

On the first line the code is like "I don't know if we have a value here, so let's null-propagate everything", and the third line is like "Let's force-unwrap this shit, I'm mostly certain it's not null". 

The null-propagation make sense if then you add a guard/if checking nullity. Now it just delaying the inevitable.

For this code to make sense you should ether:

- Change the `as?` to `as!` (and it blow up in case something is null, that's what will happen with the current code anyway, but at least the operators will be consistent. Maybe Bundle.main.saraza is never null so it's OK to do `as!`)

- Add an if/guard check complementary to `as?` operator and then act accordingly if domain is null.